docs/source/designs/backup-recovery.rst

```diff
@@ -0,0 +1,515 @@
+
+Sample Sessions
+~~~~~~~~~~~~~~~
+
+The expected interaction pattern involves two API calls.
+
+#. Early in the Tahoe-LAFS client node setup/configuration process,
+   configure replication:
+
+   .. code-block:: html
+
+      POST /storage-plugins/privatestorageio-zkapauthz-v1/replicate
+
+      201 Created
+      Content-Type: application/json
+
+      {"recovery-capability": "URI:DIR-RO:xxxx"}
+
+#. Normal use of the Tahoe-LAFS client node,
+   including redeeming vouchers and spending ZKAPs.
+
+#. After losing the Tahoe-LAFS client node,
+   create a new Tahoe-LAFS client node and recover from the replica:
+
+   .. code-block:: html
+
+      POST /storage-plugins/privatestorageio-zkapauthz-v1/recover
+      Content-Type: application/json
+      Content-Length: ...
+
+      {"recovery-capability": "URI:DIR-RO:xxxx"}
+
+      200 OK
+      Content-Type: application/json
+
+      {}
+
+#. The new Tahoe-LAFS client node now has the same ZKAPAuthorizer state as it did prior to lose of the original instance.
+
+Data Integrity
+--------------
+
+Schema Upgrades
+~~~~~~~~~~~~~~~
+
+A database snapshot will include schema modification statements
+(DDL statements)
+which completely initialize the schema for all subsequent data manipulation statements
+(DML statements)
+in the snapshot.
+
+An event stream must contain information about schema modifications because different statements in the stream may require different versions of the schema.
+This will happen whenever
+
+#. a snapshot is created
+#. some statements are recorded in the event stream
+#. a schema upgrade is performed (e.g. as a result of client software upgrade)
+#. more statements are recorded in the event stream
+
+These requirements can be exactly satisfied if DDL and DML statements are handled uniformly.
+If DDL statements are recorded in the event stream and later executed during recovery the schema will always match the requirements of the DML statements.
+
+Automated Testing
+~~~~~~~~~~~~~~~~~
+
+The replication/recovery functionality can be implemented orthogonally to ZKAPAuthorizer application logic.
+This means it can be tested orthogonally to ZKAPAuthorizer application logic.
+This means the core logic should be amenable to high-quality unit testing.
+
+Successful replication in practice depends on reads from and writes to Tahoe-LAFS storage.
+Automated testing for this logic probably requires integration-style testing due to the lack of unit testing affordances from the Tahoe-LAFS project.
+
+Runtime Health-Checks
+~~~~~~~~~~~~~~~~~~~~~
+
+The maintenace of a replica is an ongoing process.
+The replica loses value,
+up to and including *all* value,
+if that maintenance process breaks down at some point.
+
+Ideally it would be possible for some component to detect problems with this process.
+Where possible,
+problems should be corrected automatically.
+At some point the system may determine no automatic correction is possible and user intervention is required.
```

Comment:
I realize that the user interaction part is out of scope here, but how/where will such errors be reported? Perhaps in the body of `GET /storage-plugins/privatestorageio-zkapauthz-v1/replicate`?

Reply:
Absent a user-interaction design that covers exposing this information, it will probably just end up in the Tahoe-LAFS client node log.